@here/harp-mapview/lib/MapViewAtmosphere.ts

```diff
@@ -0,0 +1,478 @@
+        // actual earth geometry.
+        let onBeforeCallback: (_camera: THREE.Camera, _material: THREE.Material) => void;
+        if (this.m_materialVariant !== AtmosphereShadingVariant.ScatteringShader) {
+            // Setup only further clip planes before rendering.
+            onBeforeCallback = (camera: THREE.Camera, _material: THREE.Material) => {
+                this.overrideClipPlanes(camera);
+            };
+        } else {
+            // Setup proper clip planes and update uniforms values.
+            onBeforeCallback = (camera: THREE.Camera, material: THREE.Material) => {
+                this.overrideClipPlanes(camera);
+                // Check material wasn't swapped.
+                assert(material instanceof SkyAtmosphereMaterial);
+                const mat = this.m_skyMaterial as SkyAtmosphereMaterial;
+                mat.updateUniforms(mat, this.m_skyMesh!, camera, this.m_lightDirection);
+            };
+        }
+
+        // Sky material should be already created with mesh.
+        assert(this.m_skyMaterial !== undefined);
+        this.m_skyMesh.onBeforeRender = (
+            _renderer: THREE.WebGLRenderer,
+            _scene: THREE.Scene,
+            camera: THREE.Camera,
+            _geometry: THREE.Geometry | THREE.BufferGeometry,
+            material: THREE.Material,
+            _group: THREE.Group
+        ) => {
+            onBeforeCallback(camera, material);
+        };
+
+        this.m_skyMesh.onAfterRender = (
+            _renderer: THREE.WebGLRenderer,
+            _scene: THREE.Scene,
+            camera: THREE.Camera,
+            _geometry: THREE.Geometry | THREE.BufferGeometry,
+            _material: THREE.Material,
+            _group: THREE.Group
+        ) => {
+            this.revertClipPlanes(camera);
+        };
+    }
+
+    private setupGroundForRendering(): void {
+        if (this.m_groundMesh === undefined) {
+            return;
+        }
+        if (this.m_materialVariant !== AtmosphereShadingVariant.ScatteringShader) {
+            return;
+        }
+        // Ground material should be already created.
+        assert(this.m_groundMaterial !== undefined);
+        // Ground mesh does not need custom clip planes and uses the same camera setup as
+        // real (data source based) geometry.
+        this.m_groundMesh.onBeforeRender = (
+            _renderer: THREE.WebGLRenderer,
+            _scene: THREE.Scene,
+            camera: THREE.Camera,
+            _geometry: THREE.Geometry | THREE.BufferGeometry,
+            material: THREE.Material,
+            _group: THREE.Group
+        ) => {
+            assert(material instanceof GroundAtmosphereMaterial);
+            const mat = this.m_groundMaterial as GroundAtmosphereMaterial;
+            mat.updateUniforms(mat, this.m_groundMesh!, camera, this.m_lightDirection);
+        };
+    }
+
+    private overrideClipPlanes(rteCamera: THREE.Camera) {
+        const { camera, projection } = this.m_mapView;
```

Comment:
This is the biggest problem, as you may see we use both "rendering" (so called relative to eye) camera as it goes to rendering pipeline and our globally positioned MapView available camera.
Please answer how could we possibly "update" MapViewAtmosphere with map view camera reference without creating MapView->MapViewAtmosphere dependency.
We may either leave MapViewAtmosphere -> MapView dependency or provide callback mechanism to update MapViewAtmosphere from MapView in generic way. My proposal is via MapView->MapAnchor that would contain optional callback method, so the interface would look like this:
```
export type MapAnchor<T extends THREE.Object3D = THREE.Object3D> = T & {
    /**
     * The position of this [[MapAnchor]] in [[GeoCoordinates]].
     */
    geoPosition?: GeoCoordinates;
    /**
     * Flag defining if the object may be picked.
     *
     * @note By default all objects are pick-able even if this flag is undefined.
     */
    pickable?: boolean;
    /**
     * Optional callback to get updates from MapView.
     */
    updateCallback?: (mapView: MapView, time: number) => void;
};
```